scripts/skills/special/rf_poise_effect.nut

```diff
@@ -0,0 +1,142 @@
+
+		return this.m.Name + suffix;
+	}
+
+	function onTurnStart()
+	{
+		this.resetPoise();
+	}
+
+	function onBeforeDamageReceived( _attacker, _skill, _hitInfo, _properties )
+	{
+		if (_skill == null || !_skill.m.IsDamagingPoise) return;
+
+		local attackPower = this.calculatePoiseDamage(_attacker, _skill, _hitInfo.BodyPart == ::Const.BodyPart.Head);
+		if (attackPower == 0) return;
+
+		this.onPoiseDamageReceived(_attacker, _skill, attackPower, true, _properties);
+	}
+
+// New Functions
+	function getPoise( _properties = null )
+	{
+		if (_properties == null) _properties = this.getContainer().getActor().getCurrentProperties();
+
+		return _properties.getPoiseMax() - this.m.InversePoise;
+	}
+
+	function resetPoise()
+	{
+		this.m.InversePoise = 0;
+	}
+
+	// Can be called manually from the outside, e.g. by charge skills that do not use attacks
+	// Returns the amount of turns that the target was stunned. Can be used to apply other effects/behaviors when called manually
+	function onPoiseDamageReceived( _attacker, _skill, _attackPower, _printLog = true, _properties = null )
+	{
+		this.m.InversePoise += _attackPower;
+
+		local turnsStunned = 0;
+		if (_skill.m.IsStunningFromPoise)
+		{
+			if (_properties == null) _properties = this.getContainer().getActor().getCurrentProperties();
+
+			turnsStunned = this.calculateTurnsStunned(_properties.getPoiseMax());
+			if (turnsStunned == 0) return;
+
+			local stunnedEffect = this.getContainer().getSkillByID("effects.stunned");
+			if (stunnedEffect == null)
+			{
+				stunnedEffect = ::new("scripts/skills/effects/stunned_effect");
+				this.getContainer().add(stunnedEffect);
+			}
+
+			stunnedEffect.setTurns(turnsStunned);
+			if (_printLog)
+			{
+				::Tactical.EventLog.log(::Const.UI.getColorizedEntityName(_attacker) + " has stunned " + ::Const.UI.getColorizedEntityName(this.getContainer().getActor()) + " for " + ::MSU.Text.colorGreen(turnsStunned) + " turn");
+			}
+
+			this.resetPoise();
+		}
+
+		return turnsStunned;
+	}
+
+	// Returns, for how long we are being stunned right now (1 parameter) or would be, if given a simulated attack power
+	function calculateTurnsStunned( _PoiseMax, _simulatedAttackPower = 0 )	// _simulatedAttackPower is only used here when simulating whether something would stun
+	{
+		if (this.m.ReversePoise + _simulatedAttackPower < _PoiseMax) return 0;
+		if (this.m.ReversePoise + _simulatedAttackPower < (_PoiseMax * 2)) return 1;
+		return 2;
+	}
+
+	function calculatePoiseDamage( _attacker, _skill, _isHeadshot)
+	{
+		if (!_skill.m.IsDamagingPoise) return 0;
+
+		local poiseDamage = _attacker.getSkills().buildPropertiesForUse(_skill, null).getPoiseDamage();
+		if (poiseDamage == 0) return;
+
+		if (_isHeadshot && !this.getContainer().hasSkill("perk.steel_brow"))
+		{
+			poiseDamage *= this.m.HeadshotMult;
+		}
+		poiseDamage = ::Math.floor(poiseDamage);
+
+		return poiseDamage;
+	}
```

Comment:
Imo this should be directly from the attacker's perspective in this effect's` onAnySkillUsed` e.g. similar to how we do in `rf_reach_effect`. Then apply the damage in `onTargetHit`.